source/common/unified_memory_allocation/include/uma/memory_pool.h

```diff
@@ -10,21 +10,41 @@
 #define UMA_MEMORY_POOL_H 1
 
 #include <uma/base.h>
-#include <uma/memory_pool_ops.h>
+#include <uma/memory_provider.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
 typedef struct uma_memory_pool_t *uma_memory_pool_handle_t;
 
+struct uma_memory_pool_ops_t;
+
+/// \brief Type of a memory provider, defines purpose of the provider.
+enum uma_memory_provider_type {
```

Comment:
bit flags?

Reply:
Or perhaps we should just have METADATA and DATA types and let user pass the same provider two times (with both DATA and METADATA) if he wants. 

Right now, nothing prevents user from doing that anyway, and we basically have two ways of marking a provider as data and metadata. Also, if in such case, we can just wrap the data provider for memory tracking and ignore the metadata one.

Reply:
I've removed the DATA_AND_METADATA type.

Reply:
The immediate reason for this API was to avoid tracking the metadata allocations yes. I don't think it will have a huge overhead so we can definitely wait with adding this.

Another reason was to allow user to pass some specialized metadata providers. However, this is an unrelated feature so I think it would be best to add it separately (in another PR) to keep this one simple.

I've removed this for now and will create another PR so we can discuss this API futher.

Reply:
> Another reason was to allow user to pass some specialized metadata providers. However, this is an unrelated feature so I think it would be best to add it separately (in another PR) to keep this one simple.

It is a good point. Introducing types for providers will help a heap manager (that requires more than 1 provider) to understand the purpose of each provider. But I agree with you: we can do it as a separate PR.